openebs-features/VOLUME_3.0.md

```diff
@@ -0,0 +1,429 @@
+designs in the later sections.
+
+[C]
+Kubernetes has been coming up with new patterns for building controllers. Controllers
+are logic that is tied to Kubernetes APIs on one hand & application's operational
+requirement on the other. It seems good to implement volume provisioning logic as 
+a K8s controller & hence reap the benefits of K8s lifecycle, non-functional 
+requirements, etc. by default. This also assumes, maya to deviate from its earlier
+philosophy of being an abstraction to various container orchestrators. In other 
+words, in this new approach, maya will be tightly coupled with Kubernetes.
+
+[D]
+Maya api server needs to cater to various volume types i.e. jiva, cstor, etc
+with ease. Any new volume type should be implemented with ease. Maya api server
+should establish a pattern that can be easily followed to add or update any volume
+types. The new approach removes volume type based logic. Instead it boils down to
+configuring yamls that suit the volume type. Hence, Maya becomes a pure pass 
+through layer without bothering about volume types. Any volume that can run inside
+a container without maya can be run with maya as well. The benefits of running
+these volumes with maya are handling of orthogonal concerns related to persistent
+storage e.g. running them inside Kubernetes with its choice of storage policies.
+
+[E]
+Declarative Code is the new norm. While declarative code has its own rough edges,
+we are seeing less of symptoms in Kubernetes world. Most possible reason behind 
+this is the way declarative yamls are mapped with Go's structs. Kubernetes avoids
+/minimises the templating stuff in these declarations and instead transforms to 
+strict types for all its Kind. It has not failed in its attempt so far considering 
+the popularity of these yamls among the developers as well as operators and testers.
+Hence, it makes sense for maya to make use of these yamls as much as possible.
+Current approach of maya is to generate these yamls/structure by putting in good
+amount of logic/code. The new approach ensures that these yamls are outside this 
+logic and can be seen, updated by an operator or her preferred tool before being
+entrusted to maya. In short the new approach provides greater control to the 
+operator who is more used to these yamls.
+
+#### Design -- Volume API/Structure
+
+```go
+type Volume struct {
+  // Name of this volume
+  Name        string
+  
+  // Capacity of this volume
+  Capacity    string
+  
+  // Properties of this volume
+  Properties  Properties
+  
+  // Policies of this volume
+  Policies    Policies
+}
+
+// Properties holds various references using which volume's properties
+// can be determined
+type Properties struct {
+  // ConfigMap is the name of the ConfigMap that holds volume's properties
+  ConfigMap   string
+  
+  // EndPoint is the url of the service that provides volume's properties
+  EndPoint    string
```

Comment:
On second thoughts, I feel EndPoint should be eliminated.
It should not be a part of maya volume controller logic.
Rather, endpoints should be pulled to get the entire set of policies & get these policies installed on the env.

NOTE: IMHO Pulling these policies can be manual or some controller other than volume controller.
NOTE: Pulling of these policies can be a job of Policy CRD controller
NOTE: Operator should have set some CronJob to build a storageclass by referring to POLICY CRD 
  - Policy CRD controller will get the latest recommend policy from an endpoint & install it
  - Policy CRD controller will then create a storageclass that refers to these new policies & install it
NOTE: EndPoints should have done due verification & provide some performance numbers of this policy that is being pulled as a latest recommendation.
  - If Policy CRD controller finds the performance numbers to be BAD than the current policy it has then it need not install these policies & storageclass

There has to be some common ground/logic w.r.t Policy CRD & Endpoint to publish policy ranks/performance.